### DISCUSSION TITLE
Cleanup

### DISCUSSION THREAD

**Discussion by larsnaesbye** [mars-sim/mars-sim#381] [Ideas]
A simple code analysis run shows a lot of potential improvements across the entire project.
Some things can be fixed quite easily, like unused imports can just be removed. 

Others can't be automatically fixed by the IDE, because it'll make guesses as to variable names that might make the code less readable.

![image](https://user-images.githubusercontent.com/7767079/126450519-2f8489a4-0155-4510-a8bc-61d086f2ed10.png)

**Comment by mokun** [comment]
> A simple code analysis run shows a lot of potential improvements across the entire project.
> Some things can be fixed quite easily, like unused imports can just be removed.
> 

Sure. May be you can let us know here in this thread starting with whatever the most critical to fix.

**Reply by larsnaesbye** [reply]
Yeah. The pull requests I made and which you merged were from this list :-)

**Comment by bevans2000** [comment]
@larsnaesbye I totally agree that the project has a lot of code waiting to fail. I've been tackling some as I've been making changes but there are still a lot. In particular there is a lot of duplicate code. I would like to add a GitFlow to regularly update the static code output in SonarCloud.

https://sonarcloud.io/dashboard?id=com.github.mars-sim%3Amars-sim

**Reply by larsnaesbye** [reply]
@bevans2000 I was in this project fairly early when the codebase was something you could more or less skim through in a weekend and start contributing to right away. Then life happened and I left and came back - now there is so much outcommented code, a lot of unused libraries and other stuff that makes it hard (for me at least) to understand what it does now.

IDEs such as IntelliJ can spot the duplicate code fast and can suggest automatic refactoring. You have to check manually that it still works, as it doesn't always do it right. But I'd say that the image I posted above could easily have half of the warnings removed without much loss of sleep :)

**Reply by larsnaesbye** [reply]
By the way, that is a cool overview of the code :)

**Reply by bevans2000** [reply]
SonarCloud is pretty cool. I'm going to hook it up to the GitHub Action build so it update on each push to the master branch.
I know what you mean, When I worked with Scott on this project in the SVN days there was less code. But definitely the project is more exciting now as then (15+ years ago ?) it was more fun.

**Reply by bevans2000** [reply]
In SonarCloud I would like to focus on getting Reliability Rating to at least B. This is normally a good measure of the code base as it covers a lot of sins.

**Reply by bevans2000** [reply]
There is also a lot of exception catches that are never rethrown or reported; basically hiding potential problems.

**Reply by larsnaesbye** [reply]
Yes, let's get reliability up. I'll focus on the IntelliJ reports and try to chip away at the reported problems.

**Reply by bevans2000** [reply]
Running SonarCloud directly from GitAction when using maven is not as straight forward as I thought. Problem is stopping it running when it's on a forked repo because we don't want those results going into the central place. Need to think or a way round it.

**Reply by larsnaesbye** [reply]
Does it run on all other branches than `master`?

**Reply by bevans2000** [reply]
O I was going to configure it just for the master branch. The problem is the repository secrets you can not use in a if clause in a Git Action. So if it runs on a forked repo without the secrets the build fails. Just need to think of a way that the action only runs on the master branch of the main repo. Must be a common  requirements so there must be a solution.
It's not a problem running it periodically as an alternative, for example once a week.

**Reply by mokun** [reply]
Sorry for my messiness, especially in the early days when I first get started. 

For sure I'm learning from you guys all the great coding patterns and how you organize things.

Now, when I looked up those 599 Eclipse warnings. 

![image](https://user-images.githubusercontent.com/1168584/126838800-075cd3c8-795c-4033-88af-2789046e7a05.png)

I found a whole bunch of them coming from the steel gauges folder that we embed as source codes (not as Maven artifact).

So, it's possible we can revisit using its maven artifact. In doing so, we would remove ~90 of these warnings.

Note that I do recall in the past when I tried using its Maven artifact, because some of the methods were internal and I needed to change how it works in order for me to incorporate that gauge or meter into mars-sim. Just so you know.

**Reply by bevans2000** [reply]
Agreed, we should use the Maven artifact and drop the code. Let's make an Issue for this problem.

I think I have found a way to do the Sonarcloud now on push to master. I'll try and sort it over the weekend if I can find time.

**Reply by larsnaesbye** [reply]
Hmm... the artifact seems old: https://mvnrepository.com/artifact/eu.hansolo/SteelSeries
August 2012 is a very long time ago. Is the main developer still working on it?

**Reply by larsnaesbye** [reply]
He's moved to JavaFX instead I think - https://mvnrepository.com/artifact/eu.hansolo/Medusa

**Reply by bevans2000** [reply]
SonarCloud analysis is called when there is a push to the master branch. It doesn't trigger on forked repos.
I have added a badge to the Readme as well.

**Reply by mokun** [reply]
> He's moved to JavaFX instead I think - https://mvnrepository.com/artifact/eu.hansolo/Medusa

@larsnaesbye 

When we were trying to rebuild mars-sim UI using JavaFX back then, we did try using Medusa. It really have a lot of nice widgets we can use. I recall we even used some for presenting the Personality Traits for settlers.

**Comment by bevans2000** [comment]
For months I've been focussing on refactoring and improving the design and performance but I'm going to change tack for a while and look at the Critical & Major bugs/CodeSmells reported by SonarCloud.

**Reply by larsnaesbye** [reply]
Good idea. Let's get some better ratings and try to get the bugs, smells and vulnerabilities count down.
